### PR TITLE
Move env.update() to Builder class

### PR DESCRIPTION
**Pull request by tk0miya**
### Feature or Bugfix
- Refactoring

### Purpose
- So far, `BuildEnvironment` has two large responsibilities. The one is storing environment data, and another is converting source files to doctrees.
- This PR moves the second responsibility to `Builder` class. The reading process is a part of builds. And it is never environment's. I thnk the `Builder` class is much appropriate than `BuildEnvironment`.

### Note

I'll also separate `env.read_doc()` method later. This is step 1 of the refactoring.
My goal is simplify `BuildEnvironment` class.

### CODE REVIEW

File: sphinx/builders/__init__.py [hunk 1]
```diff
@@ -403,6 +405,109 @@ def build(self, docnames, summary=None, method='update'):
...
+        # type: () -> List[unicode]
+        """(Re-)read all files new or changed since last update.
+
+        Store all environment docnames in the canonical format (ie using SEP as
+        a separator in place of os.path.sep).
+        """
+        changed, reason = self.env.update_config(self.config, self.srcdir, self.doctreedir)
+
+        # this cache also needs to be updated every time
+        self.env._nitpick_ignore = set(self.config.nitpick_ignore)
```
Comment (tk0miya) [inline sphinx/builders/__init__.py]: It seems this "cache" is not needed. I'll remove this in other PR.

File: sphinx/builders/__init__.py [hunk 2]
```diff
@@ -403,6 +405,109 @@ def build(self, docnames, summary=None, method='update'):
...
+
+        logger.info(bold('updating environment: '), nonl=True)
+
+        self.env.find_files(self.config, self)
+        added, changed, removed = self.env.get_outdated_files(changed)
+
+        # allow user intervention as well
+        for docs in self.app.emit('env-get-outdated', self, added, changed, removed):
```
Comment (tk0miya) [inline sphinx/builders/__init__.py]: I feel `env-*` events are not good name after moving. But they are used widely by 3rd parties.
So I'm wondering about renaming.

File: sphinx/builders/__init__.py [hunk 3]
```diff
@@ -403,6 +405,109 @@ def build(self, docnames, summary=None, method='update'):
...
+        return sorted(docnames)
+
+    def _read_serial(self, docnames):
+        # type: (List[unicode]) -> None
+        for docname in status_iterator(docnames, 'reading sources... ', "purple",
+                                       len(docnames), self.app.verbosity):
+            # remove all inventory entries for that file
+            self.app.emit('env-purge-doc', self.env, docname)
+            self.env.clear_doc(docname)
+            self.env.read_doc(docname, self.app)
```
Comment (tk0miya) [inline sphinx/builders/__init__.py:475]: As I commented first, I will work about `env.read_doc()` in next PR. Stay tuned :-)

**Comment by tk0miya** [comment]
I noticed I have to update tests more. `tests/test_environment.py` emits deprecation warnings.

File: sphinx/builders/__init__.py [hunk 4]
```diff
@@ -344,7 +346,7 @@ def build(self, docnames, summary=None, method='update'):
 
         # while reading, collect all warnings from docutils
         with logging.pending_warnings():
-            updated_docnames = set(self.env.update(self.config, self.srcdir, self.doctreedir))
+            updated_docnames = set(self.read())
```
Comment (shimizukawa) [inline sphinx/builders/__init__.py:349]: 👍 👍 👍 👍 👍

**Review by shimizukawa — APPROVED** [review]
LGTM! Great work!!🎉 🎉 🎉

**Comment by tk0miya** [comment]
Thank you for quick review! I'll merge this after a while.

**Comment by tk0miya** [comment]
Merged. Thank you for reviewing!

**Pull request merged**